en/haskell-rust-ffi.md

```diff
@@ -0,0 +1,233 @@
+        ./hello-haskell
+      ];
+    };
+}
+```
+
+Stage the changes:
+
+```sh
+git add hello-haskell
+```
+
+Now, you can run `nix run .#hello-haskell` to build and execute the Haskell project.
+
+{#merge-devshell}
+## Merge Rust and Haskell Development Environments
+
+In the previous section, we created `devShells.haskell`. Let's merge it with the Rust development environment in `flake.nix`:
+
+```nix
+{
+  # Inside devShells.default
+  inputsFrom = [
+    # ...
+    self'.devShells.haskell
+  ];
+}
+```
+
+Now, re-enter the shell, and you'll have both Rust and Haskell development environments:
+
+```sh
+exit
+nix develop
+cd hello-haskell && cabal build
+cd .. && cargo build
+```
+
+{#add-rust-lib}
+## Add Rust Library as a Dependency
+
+Just like any other dependency, you'll first add it to your `.cabal` file:
+
+```cabal
+executable hello-haskell
+  -- ...
+  extra-libraries: rust_nix_template
+```
+
+Try building it:
+
+```sh
+cd hello-haskell && cabal build
+```
+
+You'll likely encounter an error like this:
+
+```sh
+...
+* Missing (or bad) C library: rust_nix_template
+...
+```
+
+The easiest solution might seem to be `export LIBRARY_PATH=../target/debug`. However, this is not reproducible and would mean running an additional command to setup the prerequisite to build the Haskell package. Even worse if the rust project is in a different repository. 
+
+Often, the easiest solution isn't the simplest. Let's use Nix to simplify this process.
+
+When you use Nix, you set up all the prerequisites beforehand, which is why you'll encounter an error when trying to re-enter the devShell without explicitly specifying where the Rust project is:
+
+```sh
+...
+error: function 'anonymous lambda' called without required argument 'rust_nix_template'
+...
+```
+
+To specify the Rust project as a dependency, let's edit `hello-haskell/default.nix` to:
+
+```nix
+{
+  # Inside haskellProjects.default
+  settings = {
+    rust_nix_template.custom = _: self'.packages.default;
+  };
+}
+```
+
+This process eliminates the need for manual Rust project building as it's wired as a prerequisite to the Haskell package.
+
+{#call-rust}
+## Call Rust function from Haskell
+
+Replace the contents of `hello-haskell/app/Main.hs` with:
+
+[[haskell-rust-ffi/hs/Main.hs]]
+![[haskell-rust-ffi/hs/Main.hs]]
+
+The implementation above is based on the [Haskell FFI documentation](https://wiki.haskell.org/Foreign_Function_Interface). Now, run the Haskell project:
+
+```sh
+nix run .#hello-haskell
+```
+
+You should see the output `Hello, from rust!`.
+
+> [!note] MacOS caveat
+> If you are on MacOS, the Haskell package will not run because `dlopen` will be looking for the `.dylib` file in the temporary build directory (`/private/tmp/nix-build-rust-nix...`). To fix this, you need to include [fixDarwinDylibNames](https://github.com/NixOS/nixpkgs/blob/af8fd52e05c81eafcfd4fb9fe7d3553b61472712/pkgs/build-support/setup-hooks/fix-darwin-dylib-names.sh) in `flake.nix`:
```

Comment:
```suggestion
> If you are on [[macos]], the Haskell package will not run because `dlopen` will be looking for the `.dylib` file in the temporary build directory (`/private/tmp/nix-build-rust-nix...`). To fix this, you need to include [fixDarwinDylibNames](https://github.com/NixOS/nixpkgs/blob/af8fd52e05c81eafcfd4fb9fe7d3553b61472712/pkgs/build-support/setup-hooks/fix-darwin-dylib-names.sh) in `flake.nix`:
```